### ISSUE TITLE
Set Time Out on Requests

### ISSUE THREAD

**Issue by imewish** [mswjs/interceptors#520]
Hello, 

Can a timeout be added to the requests on the interceptor? 

The goal is to add a timeout for any intercepted request and modify the request.

**Comment by kettanaito** [comment]
Hi, @imewish. Can you give me an example of what you'd use that timeout for? To cancel a request, I presume? Would love to learn more.

**Comment by imewish** [comment]
Hey @kettanaito , yes I want to cancel a request if it's taking more than X Milliseconds. Similar to timeout specified in AXIOS.

**Comment by kettanaito** [comment]
This is something I'd love to add too. This is the API I was thinking about, what's your opinion on it?

```js
interceptor.on('request', ({ request, abortController }) => {
  // Abort this intercepted request if it's taking more than 1s.
  setTimeout(() => abortController.abort(), 1000)
})
```

The problem with this requirement is this: what defines the "taking more than X"? Is it X amount of time as the request is being sent? Before the first byte of the response is received? Before the entire response is received (cannot really cancel afterward)? 

Let me know how you'd expect this to behave.

**Comment by imewish** [comment]
Hey @kettanaito , So ideally one should be able to cancel any requests in both cases you mentioned. Im not sure if this is handled differently incase of fetch/axios when a user specifies the timeout params.

**Comment by kettanaito** [comment]
> eg, if the client was not able to make the network connection to the target domain

If you don't respond with a mock, the request will happen as-is. The connection shouldn't hang in either case. 

> Also if the connection was made successfully and the response is taking too much of time we should be able to cancel it as well.

My main question is: how do you describe this in the listener? You don't know which stage of the resolution the request is. The listener is called as soon as the request headers are sent. So you can abort while the request body is streaming, once it's done streaming, once it's waiting for the response. The problem is, you cannot differentiate between any of those in the listener because there's no way of knowing that. 

Do you see my concern here? For you, the user of the library, it's a bad experience.

**Comment by imewish** [comment]
Yeah, I got your concern. So if we cannot differentiate b/w request and response timeout, a generic implementation like you have mentioned here https://github.com/mswjs/interceptors/issues/520#issuecomment-2002618245 should be ok. 

So the main use case for me here is I want to abort all the HTTP calls that are used in the code behind my APIs, the API calls we make using axios or fetch, we can set our own timeout config with axios/fetch instances. Still, the calls are made behind the scenes of other libraries/SDKs we don't have control. When we were experiencing high latencies in our APIs, and had no clue what was causing the delay, using this library I was able to trace most of the calls made by these libraries. So this thought came in, if we can trace/log these calls, if we can add timeout/abort it will be very useful.
